## Using Linear Regression algorithm

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import sklearn.datasets as data

In [2]:
df=pd.read_csv("Downloads/Cleaned_Bengaluru_House_Data.csv")
df

bath  balcony   price  total_sqft_num  size_bhk  price_per_sqft  \
0      3.0      2.0  150.00          1672.0         3     8971.291866   
1      3.0      3.0  149.00          1750.0         3     8514.285714   
2      3.0      2.0  150.00          1750.0         3     8571.428571   
3      2.0      2.0   40.00          1250.0         2     3200.000000   
4      2.0      2.0   83.00          1200.0         2     6916.666667   
...    ...      ...     ...             ...       ...             ...   
7115   3.0      2.0  325.00          2900.0         3    11206.896552   
7116   3.0      1.0   84.83          1780.0         3     4765.730337   
7117   2.0      1.0   48.00           880.0         2     5454.545455   
7118   2.0      1.0   55.00          1000.0         2     5500.000000   
7119   2.0      1.0   78.00          1400.0         3     5571.428571   

      area_typeSuper built-up  Area  area_typeBuilt-up  Area  \
0                                 1                        0   
1                                 0                        1   
2                                 1                        0   
3                                 1                        0   
4                                 0                        0   
...                             ...                      ...   
7115                              1                        0   
7116                              1                        0   
7117                              0                        0   
7118                              0                        0   
7119                              0                        0   

      area_typePlot  Area  availability_Ready To Move  ...  \
0                       0                           1  ...   
1                       0                           1  ...   
2                       0                           1  ...   
3                       0                           1  ...   
4                       1                           1  ...   
...                   ...                         ...  ...   
7115                    0                           1  ...   
7116                    0                           0  ...   
7117                    1                           1  ...   
7118                    1                           0  ...   
7119                    1                           0  ...   

      location_Kalena Agrahara  location_Horamavu Agara  \
0                            0                        0   
1                            0                        0   
2                            0                        0   
3                            0                        0   
4                            0                        0   
...                        ...                      ...   
7115                         0                        0   
7116                         0                        0   
7117                         0                        0   
7118                         0                        0   
7119                         0                        0   

      location_Vidyaranyapura  location_Hosur Road  location_Hebbal Kempapura  \
0                           0                    0                          0   
1                           0                    0                          0   
2                           0                    0                          0   
3                           0                    0                          0   
4                           0                    0                          0   
...                       ...                  ...                        ...   
7115                        0                    0                          0   
7116                        0                    0                          0   
7117                        0                    0                          0   
7118                        0                    0                          0   
7119           

In [3]:
x=df.drop("price",axis=1)
y=df["price"]

In [4]:
x

bath  balcony  total_sqft_num  size_bhk  price_per_sqft  \
0      3.0      2.0          1672.0         3     8971.291866   
1      3.0      3.0          1750.0         3     8514.285714   
2      3.0      2.0          1750.0         3     8571.428571   
3      2.0      2.0          1250.0         2     3200.000000   
4      2.0      2.0          1200.0         2     6916.666667   
...    ...      ...             ...       ...             ...   
7115   3.0      2.0          2900.0         3    11206.896552   
7116   3.0      1.0          1780.0         3     4765.730337   
7117   2.0      1.0           880.0         2     5454.545455   
7118   2.0      1.0          1000.0         2     5500.000000   
7119   2.0      1.0          1400.0         3     5571.428571   

      area_typeSuper built-up  Area  area_typeBuilt-up  Area  \
0                                 1                        0   
1                                 0                        1   
2                                 1                        0   
3                                 1                        0   
4                                 0                        0   
...                             ...                      ...   
7115                              1                        0   
7116                              1                        0   
7117                              0                        0   
7118                              0                        0   
7119                              0                        0   

      area_typePlot  Area  availability_Ready To Move  location_Whitefield  \
0                       0                           1                    0   
1                       0                           1                    0   
2                       0                           1                    0   
3                       0                           1                    0   
4                       1                           1                    0   
...                   ...                         ...                  ...   
7115                    0                           1                    0   
7116                    0                           0                    0   
7117                    1                           1                    0   
7118                    1                           0                    0   
7119                    1                           0                    0   

      ...  location_Kalena Agrahara  location_Horamavu Agara  \
0     ...                         0                        0   
1     ...                         0                        0   
2     ...                         0                        0   
3     ...                         0                        0   
4     ...                         0                        0   
...   ...                       ...                      ...   
7115  ...                         0                        0   
7116  ...                         0                        0   
7117  ...                         0                        0   
7118  ...                         0                        0   
7119  ...                         0                        0   

      location_Vidyaranyapura  location_Hosur Road  location_Hebbal Kempapura  \
0                           0                    0                          0   
1                           0                    0                          0   
2                           0                    0                          0   
3                           0                    0                          0   
4                           0                    0                          0   
...                       ...                  ...                        ...   
7115                        0                    0                          0   
7116                        0                    0                          0   
7117                 

In [5]:
y

0       150.00
1       149.00
2       150.00
3        40.00
4        83.00
         ...  
7115    325.00
7116     84.83
7117     48.00
7118     55.00
7119     78.00
Name: price, Length: 7120, dtype: float64

In [6]:
from sklearn.model_selection import train_test_split 

In [7]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=51)
print(x_train.shape,x_test.shape,y_train.shape,y_test.shape)

(5696, 107) (1424, 107) (5696,) (1424,)


Above we have splitted the **input dataset (x)** into **x_train & x_test dataset** and **output dataset (y)** into **y_train & y_test dataset**.

In [8]:
x_train

bath  balcony  total_sqft_num  size_bhk  price_per_sqft  \
7112   5.0      2.0          2500.0         6     7400.000000   
2169   2.0      0.0           995.0         2     4623.115578   
47     2.0      2.0          1180.0         2     7500.000000   
446    2.0      1.0          1450.0         3     5862.068966   
2650   2.0      1.0          1197.0         2     6432.748538   
...    ...      ...             ...       ...             ...   
6672   2.0      1.0          1246.0         2     5457.463884   
1733   2.0      1.0          1010.0         2     5742.574257   
1760   2.0      2.0          1080.0         2     4629.629630   
485    2.0      2.0           982.0         2     2500.000000   
2105   2.0      1.0          1000.0         2     2888.000000   

      area_typeSuper built-up  Area  area_typeBuilt-up  Area  \
7112                              0                        1   
2169                              0                        1   
47                                1                        0   
446                               1                        0   
2650                              1                        0   
...                             ...                      ...   
6672                              1                        0   
1733                              0                        1   
1760                              1                        0   
485                               1                        0   
2105                              1                        0   

      area_typePlot  Area  availability_Ready To Move  location_Whitefield  \
7112                    0                           1                    0   
2169                    0                           1                    0   
47                      0                           1                    0   
446                     0                           1                    0   
2650                    0                           1                    0   
...                   ...                         ...                  ...   
6672                    0                           1                    0   
1733                    0                           1                    0   
1760                    0                           1                    0   
485                     0                           0                    0   
2105                    0                           0                    0   

      ...  location_Kalena Agrahara  location_Horamavu Agara  \
7112  ...                         0                        0   
2169  ...                         0                        0   
47    ...                         0                        0   
446   ...                         0                        0   
2650  ...                         0                        0   
...   ...                       ...                      ...   
6672  ...                         0                        0   
1733  ...                         0                        0   
1760  ...                         0                        0   
485   ...                         0                        0   
2105  ...                         0                        0   

      location_Vidyaranyapura  location_Hosur Road  location_Hebbal Kempapura  \
7112                        0                    0                          0   
2169                        0                    0                          0   
47                          0                    0                          0   
446                         0                    0                          0   
2650                        0                    0                          0   
...                       ...                  ...                        ...   
6672                        0                    0                          0   
1733                        0                    0                          0   
1760                 

In [9]:
from sklearn.preprocessing import StandardScaler

We have applied the **Standardization** technique on **x_train & x_test** dataset using **StandardScaler** function.**Standardization** technique scales the dataset between **-3 to +3** and returns the 2D array. you can see below.

In [10]:
sc=StandardScaler()
x_train_SSc=sc.fit_transform(x_train)
x_test_SSc=sc.fit_transform(x_test)

In [11]:
x_train_SSc

array([[ 3.0245515 ,  0.55008644,  1.33396246, ..., -0.05471275,
        -0.05138463, -0.05630391],
       [-0.44274597, -2.0427027 , -0.60737352, ..., -0.05471275,
        -0.05138463, -0.05630391],
       [-0.44274597,  0.55008644, -0.36873754, ..., -0.05471275,
        -0.05138463, -0.05630391],
       ...,
       [-0.44274597,  0.55008644, -0.49772996, ..., -0.05471275,
        -0.05138463, -0.05630391],
       [-0.44274597,  0.55008644, -0.62414253, ..., -0.05471275,
        -0.05138463, -0.05630391],
       [-0.44274597, -0.74630813, -0.6009239 , ..., -0.05471275,
        -0.05138463, -0.05630391]])

In [12]:
x_test_SSc

array([[ 0.62766611,  0.03037637,  0.12398491, ..., -0.04594768,
        -0.05935999, -0.03750293],
       [-0.45835641,  0.57222965, -0.39064285, ..., -0.04594768,
        -0.05935999, -0.03750293],
       [ 1.71368862,  0.57222965,  0.31207985, ..., -0.04594768,
        -0.05935999, -0.03750293],
       ...,
       [-1.54437893,  0.57222965, -0.65472815, ..., -0.04594768,
        -0.05935999, -0.03750293],
       [-0.45835641, -0.7314798 , -0.27703351, ..., -0.04594768,
        -0.05935999, -0.03750293],
       [-0.45835641, -2.03518925,  0.02316602, ..., -0.04594768,
        -0.05935999, -0.03750293]])

In [22]:
from sklearn.linear_model import LinearRegression 

**Creating and applying linear regression model**

In [24]:
lr=LinearRegression ()
lr.fit(x_train_SSc,y_train)

LinearRegression()

**regression coefficients ($\beta_1,\beta_2,\beta_3,.......$)**

In [25]:
lr.coef_

array([-5.70206143e+00, -1.25679916e+00,  8.27341833e+01, -1.44906911e+01,
        5.75662723e+01,  1.88468905e-01, -1.72593897e+00, -4.51058311e+00,
       -2.22589244e+00, -4.28978455e+00, -2.44590976e+00,  5.40246226e-01,
       -1.03633400e+00,  1.43064873e+00, -6.25029424e-02, -1.51548783e+00,
       -2.14422789e-01,  2.16244155e+00, -1.48710228e+00,  1.95250816e+00,
       -3.10761125e+00, -1.01367155e+00, -1.28138668e+00,  1.37968545e-02,
        1.10383858e+00,  1.26497611e+00, -3.52405517e+00, -5.04622019e-01,
       -1.21398741e+00,  1.46299181e+00, -5.50064233e-01, -8.46468162e-02,
        6.84882188e-01, -1.39849820e+00, -1.94761710e-02,  4.20886278e-01,
       -1.57716300e+00,  8.03443207e-01,  3.86430413e-03,  1.05037261e-01,
        2.89115612e-01,  2.99182164e+00, -3.10533604e+00,  1.05625868e+00,
       -1.39649279e+00, -3.16916626e-01, -7.49672917e-02,  1.01764011e-01,
       -8.03271555e-01, -1.27061856e+00, -8.54046164e-01, -9.07215234e-01,
        9.10688839e-01,  

In [26]:
lr.intercept_

95.0802729985955

In [27]:
x_test_SSc[0,:]

array([ 0.62766611,  0.03037637,  0.12398491,  0.55931165, -0.4616068 ,
       -1.72237447,  2.12588331, -0.26736177,  0.51662744, -0.18677184,
       -0.15639831, -0.16332875, -0.12234347, -0.13093073, -0.13902205,
       -0.12234347, -0.09964476, -0.11935248, -0.12526719, -0.09964476,
       -0.116289  , -0.12526719, -0.116289  , -0.09964476, -0.09218776,
       -0.09964476, -0.08409599, -0.08409599, -0.07028523, -0.05935999,
       -0.06504853, -0.08823182, -0.07975227, -0.08823182, -0.08823182,
       -0.08823182, -0.05935999, -0.05935999, -0.07028523, -0.09598606,
       -0.03750293, -0.08409599, -0.04594768, -0.04594768, -0.06504853,
       -0.07975227, -0.05935999, -0.09598606, -0.08409599, -0.04594768,
       -0.06504853, -0.08409599, -0.04594768, -0.05935999, -0.05935999,
       -0.08409599, -0.08409599, -0.06504853, -0.07028523, -0.06504853,
       -0.05307449, -0.05307449, -0.0751646 , -0.07028523, -0.07975227,
       -0.04594768, -0.05935999, -0.0751646 , -0.05935999, -0.07

**Making Prediction**

In [18]:
lr.predict([x_test_SSc[0,:]])

array([65.02298605])

In [29]:
pred=lr.predict(x_test_SSc)
pred

array([ 65.02298605,  33.23090063,  93.96397969, ...,  51.89571943,
        80.46695654, 212.64731324])

In [30]:
y_test

2435     80.00
3113     40.00
426     120.00
1124     79.00
1161     45.00
         ...  
2078     28.34
6855     84.00
4381     32.00
3862     63.00
43      180.00
Name: price, Length: 1424, dtype: float64

**R2 score on test and train dataset**

In [32]:
lr.score(x_test_SSc,y_test)

0.8025717194735488

In [33]:
lr.score(x_train_SSc,y_train)

0.8655498398662171

## Using regularization techniques
### (i) Ridge Rgression

In [34]:
from sklearn.linear_model import Ridge, Lasso

In [35]:
rd=Ridge()
rd.fit(x_train_SSc,y_train)

Ridge()

In [25]:
rd.coef_

array([-5.69057569e+00, -1.25625139e+00,  8.26974090e+01, -1.44705656e+01,
        5.75539464e+01,  2.14173881e-01, -1.70241460e+00, -4.48971310e+00,
       -2.22556087e+00, -4.28683207e+00, -2.44483502e+00,  5.39422705e-01,
       -1.03584911e+00,  1.42947915e+00, -6.27620370e-02, -1.51509287e+00,
       -2.14597105e-01,  2.16055324e+00, -1.48668111e+00,  1.95092147e+00,
       -3.10536408e+00, -1.01343993e+00, -1.28104538e+00,  1.34714202e-02,
        1.10302840e+00,  1.26385822e+00, -3.51891843e+00, -5.04369455e-01,
       -1.21299723e+00,  1.46174005e+00, -5.49945140e-01, -8.50105063e-02,
        6.84441105e-01, -1.39776488e+00, -1.99918700e-02,  4.20587117e-01,
       -1.57619396e+00,  8.02475179e-01,  3.80130289e-03,  1.05015182e-01,
        2.88437621e-01,  2.99018469e+00, -3.10172418e+00,  1.05550113e+00,
       -1.39586504e+00, -3.16753582e-01, -7.51655997e-02,  1.01812858e-01,
       -8.02828310e-01, -1.26669906e+00, -8.54288673e-01, -9.06896404e-01,
        9.09683910e-01,  

In [26]:
rd.intercept_

95.0802729985955

In [30]:
rd.alpha

1.0

In [36]:
rd.score(x_test_SSc,y_test)

0.802543990523332

In [37]:
rd.score(x_train_SSc,y_train)

0.8655497750101236

In [36]:
rd1=Ridge(alpha=2)
rd1.fit(x_train_SSc,y_train)
rd1.score(x_test_SSc,y_test)

0.8025161794370435

## (ii) Lasso Regression

In [38]:
ls=Lasso()
ls.fit(x_train_SSc,y_train)

Lasso()

In [33]:
ls.coef_

array([-4.87715080e+00, -4.12521656e-01,  7.83306471e+01, -1.16600598e+01,
        5.22157086e+01,  7.87456282e-01, -0.00000000e+00, -1.03701845e+00,
       -7.12795470e-01, -2.41520649e+00, -8.69644061e-01,  0.00000000e+00,
       -0.00000000e+00,  1.94973515e-01,  0.00000000e+00, -1.08343110e-01,
        0.00000000e+00,  8.74376134e-01, -6.53521245e-02,  8.13068622e-01,
       -1.19618840e+00, -0.00000000e+00, -0.00000000e+00,  0.00000000e+00,
        0.00000000e+00,  2.00297179e-02, -7.59886114e-01, -0.00000000e+00,
       -0.00000000e+00,  1.71015410e-01, -0.00000000e+00, -0.00000000e+00,
        0.00000000e+00, -6.48031069e-02, -0.00000000e+00,  0.00000000e+00,
       -1.44617602e-01,  0.00000000e+00, -0.00000000e+00,  0.00000000e+00,
        0.00000000e+00,  1.82349037e+00, -1.03393703e+00,  0.00000000e+00,
       -6.52569355e-02, -0.00000000e+00, -0.00000000e+00,  0.00000000e+00,
       -0.00000000e+00,  0.00000000e+00, -0.00000000e+00, -0.00000000e+00,
        0.00000000e+00,  

In [34]:
ls.intercept_

95.08027299859552

In [39]:
ls.score(x_test_SSc,y_test)

0.7936228492392652

In [40]:
ls1=Lasso(alpha=2)
ls1.fit(x_train_SSc,y_train)
ls1.score(x_test_SSc,y_test)

0.7873577097747054

In **Ridge & Lasso Regression** **alpha ($\alpha$)** is the parameter which balances the amount of emphasis given to minimizing RSS vs minimizing sum of square of coefficients.

i.e if $\alpha=0$:  The objective become same as linear regression, i.e. we will get the same ccoefficient as simple linear regression.

if **$\alpha=\infty$**:  The coefficients will be zero.

if **$0<\alpha<\infty$**: The magnitude of $\alpha$ will decide the weightage given to different parts of objective.

from above regression algorithms with regularization techniques **linear regression & Ridge regression** performs same based on **r2 score** but **Lasso regression** underperformed as campared to **Linear and Ridge**.

For more better performance lets use some more algorithms.

## Using Polynomial Regression

In [43]:
from sklearn.linear_model import LinearRegression 
from sklearn.preprocessing import PolynomialFeatures

To apply **Ploynomial Regression** on a certain dataset, first we have to convert that dataset into polynomial. Using **PolynomialFeatures** function we can convert dataset into polynomial. You can see below the procedure.

In [44]:
poly_reg=PolynomialFeatures(degree=2)
x_train_SSc_poly=poly_reg.fit_transform(x_train_SSc)
x_test_SSc_poly=poly_reg.fit_transform(x_test_SSc)
print(x_train_SSc_poly.shape,x_test_SSc_poly.shape)

(5696, 5886) (1424, 5886)


**Applying Linear regression on plynomial dataset**

In [45]:
poly_lr=LinearRegression ()
poly_lr.fit(x_train_SSc_poly,y_train)

LinearRegression()

In [46]:
poly_lr.coef_

array([14.37726708, -0.13881672, -0.11444317, ...,  0.1206288 ,
        0.03150604,  0.05968657])

In [47]:
poly_lr.intercept_

83.9689143273028

In [48]:
y_predict1=poly_lr.predict(x_test_SSc_poly)
y_predict1

array([ 73.80089703,  59.05005159,  90.49010757, ...,  45.94398641,
        66.64207024, 167.31867642])

**r2 score**

In [49]:
poly_lr.score(x_test_SSc_poly,y_test)

0.8427598030730575

In [22]:
y_test

2435     80.00
3113     40.00
426     120.00
1124     79.00
1161     45.00
         ...  
2078     28.34
6855     84.00
4381     32.00
3862     63.00
43      180.00
Name: price, Length: 1424, dtype: float64

As compared to ridge and linear **Polynomial** gives better performance.

## Using Decision Tree 

We can also solve regression problems using **Decision Tree** algorithm by applying **DescisionTreeRegressor** function.

In [56]:
from sklearn.tree import DecisionTreeRegressor

**Applying on scaled data**

In [124]:
regression=DecisionTreeRegressor(criterion= "squared_error")
regression.fit(x_train_SSc,y_train)

DecisionTreeRegressor()

In [125]:
regression.score(x_test_SSc,y_test)

0.751192077457934

**Applying on unscaled data**

In [126]:
regression1=DecisionTreeRegressor(criterion= "squared_error")
regression1.fit(x_train,y_train)

DecisionTreeRegressor()

In [127]:
regression1.score(x_test,y_test)

0.867195298610535

Decision Tree gives very good performance on unscaled data

## Using Random Forest 

In [83]:
from sklearn.ensemble import RandomForestRegressor 

**on scaled data**

In [128]:
RF_regression=RandomForestRegressor(n_estimators=100,criterion="squared_error")
RF_regression.fit(x_train_SSc,y_train)

RandomForestRegressor()

In [129]:
RF_regression.score(x_test_SSc,y_test)

0.7413760291725486

**On unscaled data**

In [94]:
RF_regression1=RandomForestRegressor(n_estimators=100,criterion="squared_error")
RF_regression1.fit(x_train,y_train)

RandomForestRegressor()

In [95]:
RF_regression1.score(x_test,y_test)

0.8981702061887786

In [130]:
RF_regression1.predict(x_test)

array([ 79.7523 ,  39.9341 , 119.98   , ...,  33.41755,  63.0763 ,
       174.69   ])

In [131]:
y_test

2435     80.00
3113     40.00
426     120.00
1124     79.00
1161     45.00
         ...  
2078     28.34
6855     84.00
4381     32.00
3862     63.00
43      180.00
Name: price, Length: 1424, dtype: float64

**sample Prediction**

In [133]:
house_inform=np.array([3.0,3.000000,2000.0,4,7500.007874,1,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
                       0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,
                       0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
                       0,0])
house_inform.shape

(107,)

In [134]:
RF_regression1.predict([house_inform])

C:\Anaconda\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but RandomForestRegressor was fitted with feature names
  warnings.warn(


array([151.33])

like Dicison tree **Random Forest** gives better performance on unscaled data.

## Using K-Nearest Neigbhors 

In [16]:
from sklearn.neighbors import KNeighborsRegressor

**On scaled data**

In [136]:
regressor=KNeighborsRegressor(n_neighbors=5)
regressor.fit(x_train_SSc,y_train)

KNeighborsRegressor()

In [137]:
regressor.score(x_test_SSc,y_test)

0.7470754379049586

**On unscaled data**

In [138]:
regressor1=KNeighborsRegressor(n_neighbors=5)
regressor1.fit(x_train,y_train)

KNeighborsRegressor()

In [139]:
regressor1.score(x_test,y_test)

0.8858214084886539

**Sample Prediction**

In [140]:
house_inform1=np.array([4.0,4.000000,5000.0,6,3000.000000,1,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
                       0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,
                       0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
                       0,0])

In [141]:
regressor1.predict([house_inform1])

C:\Anaconda\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but KNeighborsRegressor was fitted with feature names
  warnings.warn(


array([193.])

### Conclusion

**1.** Tree based algorithms like **Decision tree & Random forest** does not require **Normalization or Standardization**.

**2.** from above all algorithms **Random Forest** has given better performance i.e. **89%**.